vm/devices/serial/serial_debugcon/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+            io,
+            tx_buffer: VecDeque::new(),
+            tx_waker: None,
+        }
+    }
+
+    /// Synchronize interrupt and waker state with device state.
+    fn sync(&mut self) {
+        // Wake to poll if there are any bytes to write.
+        if !self.tx_buffer.is_empty() {
+            if let Some(waker) = self.tx_waker.take() {
+                waker.wake();
+            }
+        }
+    }
+
+    fn poll_tx(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        while !self.tx_buffer.is_empty() {
+            let (buf, _) = self.tx_buffer.as_slices();
+            match ready!(Pin::new(&mut self.io).poll_write(cx, buf)) {
+                Ok(n) => {
+                    assert_ne!(n, 0);
+                    self.tx_buffer.drain(..n);
+                }
+                Err(err) if err.kind() == ErrorKind::BrokenPipe => {
+                    tracing::info!("debugcon output broken pipe");
+                }
+                Err(err) => {
+                    tracelimit::error_ratelimited!(
+                        len = buf.len(),
+                        error = &err as &dyn std::error::Error,
+                        "debugcon write failed, dropping data"
+                    );
+                    self.tx_buffer.drain(..buf.len());
+                }
+            }
+        }
+        // Wait for more bytes to write.
+        self.tx_waker = Some(cx.waker().clone());
+        Poll::Pending
+    }
+}
+
+impl ChangeDeviceState for SerialDebugcon {
+    fn start(&mut self) {}
+
+    async fn stop(&mut self) {}
+
+    async fn reset(&mut self) {
+        self.tx_buffer.clear();
+    }
+}
+
+impl ChipsetDevice for SerialDebugcon {
+    fn supports_pio(&mut self) -> Option<&mut dyn PortIoIntercept> {
+        Some(self)
+    }
+
+    fn supports_poll_device(&mut self) -> Option<&mut dyn PollDevice> {
+        Some(self)
+    }
+}
+
+impl PollDevice for SerialDebugcon {
+    fn poll_device(&mut self, cx: &mut Context<'_>) {
+        let _ = self.poll_tx(cx);
+        self.sync();
+    }
+}
+
+impl PortIoIntercept for SerialDebugcon {
+    fn io_read(&mut self, io_port: u16, data: &mut [u8]) -> IoResult {
+        if io_port != self.io_port {
+            return IoResult::Err(IoError::InvalidRegister);
+        }
+
+        if data.len() != 1 {
+            return IoResult::Err(IoError::InvalidAccessSize);
+        }
+
+        // this is a magic constant which the guest can use to detect the
+        // presence of the debugcon device. Its value is fixed, and matches the
+        // value set by the original debugcon implementation in bochs.
+        data[0] = 0xe9;
+
+        IoResult::Ok
+    }
+
+    fn io_write(&mut self, io_port: u16, data: &[u8]) -> IoResult {
+        if io_port != self.io_port {
+            return IoResult::Err(IoError::InvalidRegister);
+        }
+
+        if data.len() != 1 {
+            return IoResult::Err(IoError::InvalidAccessSize);
+        }
+
+        self.tx_buffer.push_back(data[0]);
```

Comment:
Instead of dropping in poll_tx, I think you want to drop here when the TX buffer exceeds some limit. Otherwise, this can grow without bound.

(Alternatively, we could stall the CPU, if we ever hit a case where the guest is usefully sending more data than we can process. But that seems unlikely.)

Reply:
Given that this is a debug device, I think the fact that the buffer grows unbounded is the desired behavior? I would wager that this isn't a device that we we would ever expect to be active in a "production" scenarios, and it might even be useful to have an infinite in-memory buffer if we wanted to, say, expose the debug buffer via `inspect`.

Reply:
Oh, I see, in poll_tx you're only dropping on error.

I still think you should put some kind of max on this, even if it's a few megabytes.

Reply:
Ok, I'll think of a number and add a cap

Reply:
done